packages/babel-parser/src/parser/lval.js

```diff
@@ -625,7 +631,8 @@ export default class LValParser extends NodeUtils {
 
     const validity = this.isValidLVal(
       expression.type,
-      hasParenthesizedAncestor || expression.extra?.parenthesized,
+      !(hasParenthesizedAncestor || expression.extra?.parenthesized) &&
+        ancestor.type === "AssignmentExpression",
```

Comment:
This code was needed to disallow `a as b = c`, however we must now make sure to _not_ disallow `[a as b] = c`.

Reply:
Can you add a new test case?
```ts
for (a as b of []);
```
It is disallowed in 7.16.2 but tsc allows it, and it seems this PR also fixes this one, too.